cmd/disk-cache.go

```
@@ -650,288 +292,90 @@ func (c cacheObjects) isCacheExclude(bucket, object string) bool {
-		default:
-		}
-		return putObjectPartFn(ctx, bucket, object, uploadID, partID, r, opts)
-	}
+// get cache disk where object is currently cached for a GET operation. If object does not exist at that location,
+// treat the list of cache drives as a circular buffer and walk through them starting at hash index
+// until an online drive is found.If object is not found, fall back to the first online cache drive
+// closest to the hash index, so that object can be re-cached.
+func (c cacheObjects) getCacheToLoc(ctx context.Context, bucket, object string) (*diskCache, error) {
+	index := c.hashIndex(bucket, object)
 
-	info = PartInfo{}
-	// Initialize pipe to stream data to backend
-	pipeReader, pipeWriter := io.Pipe()
-	hashReader, err := hash.NewReader(pipeReader, size, data.MD5HexString(), data.SHA256HexString(), data.ActualSize(), globalCLIContext.StrictS3Compat)
-	if err != nil {
-		return
-	}
-	// Initialize pipe to stream data to cache
-	rPipe, wPipe := io.Pipe()
-	cHashReader, err := hash.NewReader(rPipe, size, data.MD5HexString(), data.SHA256HexString(), data.ActualSize(), globalCLIContext.StrictS3Compat)
-	if err != nil {
-		return
-	}
-	pinfoCh := make(chan PartInfo)
-	errorCh := make(chan error)
-	go func() {
-		info, err = putObjectPartFn(ctx, bucket, object, uploadID, partID, NewPutObjReader(hashReader, nil, nil), opts)
-		if err != nil {
-			close(pinfoCh)
-			pipeWriter.CloseWithError(err)
-			wPipe.CloseWithError(err)
-			errorCh <- err
-			return
+	numDisks := len(c.cache)
+	// save first online cache disk closest to the hint index
+	var firstOnlineDisk *diskCache
+	for k := 0; k < numDisks; k++ {
+		i := (index + k) % numDisks
+		if c.cache[i] == nil {
+			continue
 		}
-		close(errorCh)
-		pinfoCh <- info
-	}()
-	go func() {
-		if _, perr := dcache.PutObjectPart(ctx, bucket, object, uploadID, partID, NewPutObjReader(cHashReader, nil, nil), opts); perr != nil {
-			wPipe.CloseWithError(perr)
-			return
+		if c.cache[i].IsOnline() {
```

This type of check does not really add any value as the disk may have gone offline just after the function call returns. It is better to simply return an error in case the disk is found to be unexpectedly offline.